### PR TITLE
PR for run query to file

### PR DESCRIPTION
**Pull request by davidkaufman**

### CODE REVIEW

**Comment by rathboma** [comment]
This looks pretty good honestly. Is it not working for all databases past 1mm records, or just certain ones?

**Comment by rathboma** [comment]
@davidkaufman whats the status here? Are you resubmitting this against ultimate?

**Comment by davidkaufman** [comment]
yes, working on it now

File: .github/workflows/studio-master.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ jobs:
       - name: Install Node.js, NPM and Yarn
         uses: actions/setup-node@v1
         with:
-          node-version: 14
+          node-version: 16
```
Comment (rathboma) [inline .github/workflows/studio-master.yml:25]: Dave your pr looks full of changes that are already in master. Not sure what’s going on, but I did see a force push. That’s likely it

Comment (davidkaufman) [reply .github/workflows/studio-master.yml]: I rebased the branch and removed these conflicts and its mergeable now ...checks are running 🤞

**Comment by davidkaufman** [comment]
hmm, the unit tests passed on my machine ...🤣

**Comment by davidkaufman** [comment]
🎉 🥳 🍻

**Comment by rathboma** [comment]
@davidkaufman is this still WIP? Still labeled as such, but I think you're done?

**Comment by davidkaufman** [comment]
Yes! @rathboma sOrry I forgot to remove the WIP label

File: apps/studio/src/components/export/ExportManager.vue [hunk 2]
```diff
@@ -59,48 +63,71 @@ export default Vue.extend({
...
           buttons: [
-            Noty.button('Show', "btn btn-primary", () => {
-              this.$native.files.showItemInFolder(options.filePath)
-              n.close()
+            Noty.button('Close', "btn btn-primary", () => {
+              error_notice.close()
             })
           ]
-        })
+        }).setTimeout(60000) // 1 minute timeout
```
Comment (rathboma) [inline apps/studio/src/components/export/ExportManager.vue:107]: Magic number! add it to globals.ts

File: apps/studio/src/components/export/ExportManager.vue [hunk 3]
```diff
@@ -59,48 +63,71 @@ export default Vue.extend({
...
+        return
+      }
+      if (exporter.status !== ExportStatus.Completed) return;
+      const n = this.$noty.success(`Export of ${exportName} complete`, {
```
Comment (rathboma) [inline apps/studio/src/components/export/ExportManager.vue:111]: not sure where it is, but if you cancel an export it says `undefined export cancelled`

File: apps/studio/src/components/export/ExportModal.vue [hunk 4]
```diff
@@ -200,9 +212,22 @@ export default {
...
       const extension = this.selectedExportFormat.key
-      return `${schema}${this.table.name}_export_${formatted}.${extension}`
+      let fileName;
+      if (this.table) {
+        const schema = this.table.schema ? `${this.table.schema}_` : ''
+        const extension = this.selectedExportFormat.key
+        fileName = `${schema}${this.table.name}_export_${formatted}.${extension}`
+      } else {
+        // sanitize query name for use as filename
+        console.log('this.query_name: ', this.query_name)
```
Comment (rathboma) [inline apps/studio/src/components/export/ExportModal.vue:223]: nitpick - `query_name` should be `queryName`.

Also -- can you remove these console.logs?

File: apps/studio/src/components/export/ExportNotification.vue [hunk 5]
```diff
@@ -27,7 +28,14 @@ export default {
   },
   computed: {
     notificationText() {
-      return `(${this.percentComplete}%) Exporting '${this.exporter.table.name}'`
+      const exportName = this.exporter.table ? this.exporter.table.name : this.exporter.query_name;
```
Comment (rathboma) [inline apps/studio/src/components/export/ExportNotification.vue:31]: same query_name vs queryName

File: apps/studio/src/components/export/ExportNotification.vue [hunk 6]
```diff
@@ -37,10 +45,13 @@ export default {
       }
       this.exporter.abort();
       this.notification.close();
-      this.$noty.error(`${this.exporter.table.name} export aborted`);
+      const exportName = this.exporter.table ? this.exporter.table.name : this.query_name;
+      this.$noty.error(`${exportName} export aborted`);
```
Comment (rathboma) [inline apps/studio/src/components/export/ExportNotification.vue:49]: I think this is where the `undefined` comes

**Review by rathboma — CHANGES_REQUESTED** [review]
This works great! Thanks for being so thorough.

The export is pretty slow though, but that's not new to your code. I think we need a file queue in the future.

Added a couple of minor comments:
- Variable name
- Notification text missing
- Looks like duplicate shortcuts in the empty query tab for run and run to file (Don't think we need a shortcut for run to file)
- Last thing - looks like the statusbar is missing from the bottom of new query tabs?

**Comment by rathboma** [comment]
@davidkaufman did you see these review comments?

**Comment by davidkaufman** [comment]
> @davidkaufman did you see these review comments?

No I missed them somehow, thanks for the nudge -- I'll check that

File: apps/studio/src/components/export/ExportManager.vue (hunk 2 above)
Comment (davidkaufman) [reply apps/studio/src/components/export/ExportManager.vue]: Fixed in 36fd713

File: apps/studio/src/components/export/ExportModal.vue (hunk 4 above)
Comment (davidkaufman) [reply apps/studio/src/components/export/ExportModal.vue]: these nits picked in 13a301a

File: apps/studio/src/components/export/ExportNotification.vue (hunk 5 above)
Comment (davidkaufman) [reply apps/studio/src/components/export/ExportNotification.vue]: also fixed in 13a301a

File: apps/studio/src/components/export/ExportNotification.vue (hunk 6 above)
Comment (davidkaufman) [reply apps/studio/src/components/export/ExportNotification.vue:49]: Good catch! this is fixed in 980017d

File: apps/studio/src/components/export/ExportManager.vue (hunk 3 above)
Comment (davidkaufman) [reply apps/studio/src/components/export/ExportManager.vue:112]: Fixed in 980017d

**Comment by davidkaufman** [comment]
@rathboma I've resolved the suggestions above, and am working on sorting out the large merge conflicts in `TabQueryEditor.vue` -- the bane of long-lived branches 😖

**Comment by rathboma** [comment]
@davidkaufman If you can resolve the merge conflict you can just merge this!

I think we need to break out code from features in future so we don't have long-lived feature branches.

Please merge asap so we don't have more conflicts.

**Comment by davidkaufman** [comment]
> @davidkaufman If you can resolve the merge conflict you can just merge this!
> 
> I think we need to break out code from features in future so we don't have long-lived feature branches.
> 
> Please merge asap so we don't have more conflicts.

Will do!

**Review by not-night-but — APPROVED** [review]
Merging away!

**Pull request merged**